apps/server/lib/lexical/server/provider/handlers/hover.ex

```diff
@@ -24,21 +30,187 @@ defmodule Lexical.Server.Provider.Handlers.Hover do
+        doc_content = module_doc(doc)
+        footer_content = module_footer(kind, module_docs)
+
+        sections = [
+          types_content,
+          doc_content,
+          footer_content
+        ]
+
+        if Enum.any?(sections, &(not empty?(&1))) do
+          {:ok, [header_content | sections]}
+        else
+          {:error, :no_doc}
+        end
+
+      _ ->
+        {:error, :no_doc}
+    end
+  end
+
+  defp hover_content({:call, module, fun, arity}, env) do
+    with {:ok, %Docs{} = module_docs} <- RemoteControl.Api.docs(env.project, module),
+         {:ok, entries} <- Map.fetch(module_docs.functions_and_macros, fun) do
+      sections =
+        entries
+        |> Enum.sort_by(& &1.arity)
+        |> Enum.filter(&(&1.arity >= arity))
+        |> Enum.flat_map(&entry_sections/1)
+
+      {:ok, sections}
+    end
+  end
+
+  defp hover_content({:type, module, type, arity}, env) do
+    with {:ok, %Docs{} = module_docs} <- RemoteControl.Api.docs(env.project, module),
+         {:ok, entries} <- Map.fetch(module_docs.types, type) do
+      case Enum.find(entries, &(&1.arity == arity)) do
+        %Docs.Entry{} = entry ->
+          {:ok, entry_sections(entry)}
+
+        _ ->
+          {:error, :no_type}
+      end
+    end
+  end
+
+  defp module_header(:module, %Docs{module: module}) do
+    module
+    |> Ast.Module.name()
+    |> Markdown.code_block()
+  end
+
+  defp module_header(:struct, %Docs{module: module}) do
+    "%#{Ast.Module.name(module)}{}"
+    |> Markdown.code_block()
+  end
+
+  defp module_doc(s) when is_binary(s), do: s
+  defp module_doc(_), do: nil
 
-      content = """
-      ### #{Ast.Module.name(module)}
+  defp module_types(:module, _), do: nil
 
-      #{doc_content}\
-      """
+  defp module_types(:struct, docs) do
+    struct_type_defs =
+      docs.types
+      |> Map.get(:t, [])
+      |> sort_entries()
+      |> Enum.flat_map(& &1.defs)
 
-      {:ok, content}
+    if struct_type_defs != [] do
+      struct_type_defs
+      |> Enum.join("\n\n")
+      |> Markdown.code_block()
+      |> Markdown.section(header: "Struct")
     end
   end
 
-  defp module_doc_content(s) when is_binary(s), do: s
-  defp module_doc_content(:none), do: "*This module is undocumented.*\n"
-  defp module_doc_content(:hidden), do: "*This module is private.*\n"
+  defp module_footer(:module, docs) do
+    callbacks = format_callbacks(docs.callbacks)
+
+    unless empty?(callbacks) do
+      Markdown.section(callbacks, header: "Callbacks")
+    end
+  end
+
+  defp module_footer(:struct, _docs), do: nil
+
+  defp entry_sections(%Docs.Entry{kind: fn_or_macro} = entry)
+       when fn_or_macro in [:function, :macro] do
+    with {:ok, call_header} <- call_header(entry) do
+      specs = Enum.map_join(entry.defs, "\n", &("@spec " <> &1))
+
+      [
+        call_header,
+        if specs != "" do
+          specs
+          |> Markdown.code_block()
+          |> Markdown.section(header: "Specs")
```

Comment:
It's largely to fix a rendering issue in VS Code hover when you have a `---` followed immediately be a code block. Since getting rid of the horizontal rules though, I'll change this too.